extra/modules/greenbids-real-time-data/src/main/java/org/prebid/server/hooks/modules/greenbids/real/time/data/config/GreenbidsRealTimeDataConfiguration.java

```diff
@@ -0,0 +1,72 @@
+package org.prebid.server.hooks.modules.greenbids.real.time.data.config;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.prebid.server.hooks.modules.greenbids.real.time.data.core.ThrottlingThresholds;
+import org.prebid.server.hooks.modules.greenbids.real.time.data.model.GreenbidsRealTimeDataProperties;
+import org.prebid.server.hooks.modules.greenbids.real.time.data.model.OnnxModelRunner;
+import org.prebid.server.hooks.modules.greenbids.real.time.data.v1.GreenbidsRealTimeDataModule;
+import org.prebid.server.hooks.modules.greenbids.real.time.data.v1.GreenbidsRealTimeDataProcessedAuctionRequestHook;
+import org.prebid.server.json.ObjectMapperProvider;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+@ConditionalOnProperty(prefix = "hooks." + GreenbidsRealTimeDataModule.CODE, name = "enabled", havingValue = "true")
+@Configuration
+public class GreenbidsRealTimeDataConfiguration {
+
+    @Bean
+    GreenbidsRealTimeDataModule greenbidsRealTimeDataModule(
+            @Value("${hooks.modules.greenbids-real-time-data.google-cloud-greenbids-project}")
+            String googleCloudGreenbidsProject,
+            @Value("${hooks.modules.greenbids-real-time-data.gcs-bucket-name}") String gcsBucketName,
+            @Value("${hooks.modules.greenbids-real-time-data.cache-expiration-minutes}") Integer cacheExpirationMinutes,
+            @Value("${hooks.modules.greenbids-real-time-data.geo-lite-country-path}") String geoLiteCountryPath,
+            @Value("${hooks.modules.greenbids-real-time-data.onnx-model-cache-key-prefix}")
+            String onnxModelCacheKeyPrefix,
+            @Value("${hooks.modules.greenbids-real-time-data.thresholds-cache-key-prefix}")
+            String thresholdsCacheKeyPrefix) {
+        final ObjectMapper mapper = ObjectMapperProvider.mapper();
+
+        final Cache<String, OnnxModelRunner> modelCacheWithExpiration = Caffeine.newBuilder()
+                .expireAfterWrite(cacheExpirationMinutes, TimeUnit.MINUTES)
+                .build();
+
+        final Cache<String, ThrottlingThresholds> thresholdsCacheWithExpiration = Caffeine.newBuilder()
+                .expireAfterWrite(cacheExpirationMinutes, TimeUnit.MINUTES)
+                .build();
+
+        final ReentrantLock lock = new ReentrantLock();
```

Comment:
Have removed the reentrant locks. Added `CompletableFuture.runAsync` to fetch from GCS in case of cache miss + added `newSingleThreadExecutor` to make only one thread access GCS (to avoid race condition)

Now the logic is as follows
- thread1 arrives -> if cache miss -> fetch from GCS -> cache.put -> return null and skip filtering
- thread2 arrives after -> if cache hit -> cache.getIfPresent -> return OnnxModelRunner and ThrottlingThresholds -> do filtering

The very first call when cache is empty is handled in the same way as cache miss -> `cache.getIfPresent` returns null and we fetch the artefacts and skip filtering after

Reply:
When the data expires in cache, I suppose the cache is invalidated and when we do `cache.getIfPresent(cacheKey)` it should return null. So it should be handled automatically by cache with expiration ?

Reply:
@EvgeniiMunin I didn't understand your take on avoiding race conditions. If you point to the cache as a shared resource for threads you shouldn't worry since that's the main point of having in-memory cache where all the fetch data is preserved until expiration. I don't see any race condition problem here. 

What I see you created a bottleneck for all the threads that would be waiting until each of them could fetch the data one by one. As I understand the GCS returns the same data for all the requests for 15 min, correct? If so and you want to minimize the number of calls to GCS you should make a synchronization/notification point where you decide whether the GCS should be called or not.

For example, you can use atomic boolean to notify all the threads that one thread starting fetching process.

Another question: does the GCS support async execution? It's worth using them.

Reply:
Added `AtomicBoolean isFetching` flag and changed executorService `newCachedThreadPool`. So now the fetch is not blocking for other threads.

By default the flag is set to false. When the thread is fetching from GCS it is switched to true with `compareAndSet`. So other threads are not blocked and skip. When the fetching is done the flag is set back to false.

Concerning GCS [doc](https://cloud.google.com/storage/docs/downloading-objects?hl=en) I don't see the functionallity specifically for async download of objects

Reply:
@EvgeniiMunin Okay, if the GCS doesn't provide any async interface then please wrap it's call with the `vertx.executeBlocking`. It will uses the vertx-based thread management of PBS. Also it's fine to operate with vertx Futures since the module eventually returns the Future.

Reply:
fixed

https://github.com/prebid/prebid-server-java/pull/3242#discussion_r1742319084